endpoints/vtrack.go

```diff
@@ -0,0 +1,332 @@
+
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Write(d)
+	}
+}
+
+/**
+ * Create vast url tracking
+ */
+func GetVastUrlTracking(externalUrl string, bidid string, bidder string, accountId string, timestamp int64) string {
+
+	eventReq := &EventRequest{
+		Type:      IMP,
+		Bidid:     bidid,
+		AccountId: accountId,
+		Bidder:    bidder,
+		Timestamp: timestamp,
+		Format:    BLANK,
+	}
+
+	return EventRequestToUrl(externalUrl, eventReq)
+}
+
+/**
+ * Parses a BidCacheRequest from an HTTP Request
+ */
+func (v *vtrackEndpoint) parseVTrackRequest(httpRequest *http.Request) (req *BidCacheRequest, err error) {
+	req = &BidCacheRequest{}
+	err = nil
+
+	// Pull the request body into a buffer, so we have it for later usage.
+	lr := &io.LimitedReader{
+		R: httpRequest.Body,
+		N: v.Cfg.MaxRequestSize,
+	}
+
+	requestJson, err := ioutil.ReadAll(lr)
+	if err != nil {
+		return req, err
+	}
+
+	// If the request size was too large, read through the rest of the request body so that the connection can be reused.
+	if lr.N <= 0 {
+		if written, err := io.Copy(ioutil.Discard, httpRequest.Body); written > 0 || err != nil {
+			err = fmt.Errorf("request size exceeded max size of %d bytes", v.Cfg.MaxRequestSize)
+			return req, err
+		}
+	}
+
+	if len(requestJson) == 0 {
+		err = fmt.Errorf("request body is empty")
+		return req, err
+	}
+
+	if err := json.Unmarshal(requestJson, req); err != nil {
+		return req, err
+	}
+
+	// validate request
+	if len(req.Puts) == 0 {
+		return req, err
+	}
+
+	for _, bcr := range req.Puts {
+		if bcr.BidID == "" {
+			err = error(&errortypes.BadInput{Message: fmt.Sprint("'bidid' is required field and can't be empty")})
+			return req, err
+		}
+
+		if bcr.Bidder == "" {
+			err = error(&errortypes.BadInput{Message: fmt.Sprint("'bidder' is required field and can't be empty")})
+			return req, err
+		}
+	}
+
+	return req, nil
+}
+
+/**
+ * Cache BidCacheRequest data
+ */
+func (v *vtrackEndpoint) cachePutObjects(req *BidCacheRequest, biddersAllowingVastUpdate []string, accountId string, timeout int64) ([]string, []error) {
+	var nputs []prebid_cache_client.Cacheable
+	sort.Strings(biddersAllowingVastUpdate)
+
+	for _, c := range req.Puts {
+
+		nc := &prebid_cache_client.Cacheable{
+			Type:       c.Type,
+			Data:       c.Data,
+			TTLSeconds: c.TTLSeconds,
+			Key:        c.Key,
+		}
+
+		if contains(biddersAllowingVastUpdate, c.Bidder) && nc.Data != nil {
+			nc.Data = modifyVastXml(v.Cfg.ExternalURL, nc.Data, c.BidID, c.Bidder, accountId, c.Timestamp)
+		}
+
+		nputs = append(nputs, *nc)
+	}
+
+	t := time.Now().Add(time.Duration(timeout) * time.Millisecond)
+	ctx, cancel := context.WithDeadline(context.Background(), t)
+	defer cancel()
+
+	return v.PbsCache.PutJson(ctx, nputs)
+}
+
+/**
+ * Returns list of bidders that allow VAST XML modification.
+ */
+func biddersAllowingVastUpdate(req *BidCacheRequest, bidderInfos *adapters.BidderInfos, allowUnknownBidder bool) []string {
```

Comment:
Consider building a `map[string]struct{}` instead to act as a hash set to get checks in O(1) instead of a O(n) search through the slice.

Reply:
Nice suggestion.